### PR TITLE
gating should not overwrite quarantine, #20531

### PR DESCRIPTION
**Pull request by patriknw**
- Reproduced with cluster sample.
  - auto-down=45s
  - node1 and node2 joins
  - suspend node2 (kill -STOP)
  - node1 removes and quarantines node2
  - resume node2 (kill -CONT)
  - node2 receives heartbeat response from node1, which
    is not expected since node2 is supposed to be quarantined
  - the reason is that the quarantine is overwritten by gating

### CODE REVIEW

**Comment by patriknw** [comment]
Refs #20531

**Comment by patriknw** [comment]
PLS BUILD

**Comment by drewhk** [comment]
LGTM

**Comment by MQ-EL** [comment]
```scala
      case NonFatal(e) ⇒
        e match {
          case _: EndpointDisassociatedException | _: EndpointAssociationException ⇒ // no logging
          case _ ⇒ log.error(e, e.getMessage)
        }
        endpoints.markAsFailed(sender(), Deadline.now + settings.RetryGateClosedFor)
        Stop
```

1. lost the refusedId if it was the ```Pass``` state before?

File: akka-remote/src/main/scala/akka/remote/Remoting.scala [hunk 1]
```diff
@@ -372,8 +372,13 @@ private[remote] object EndpointManager {
      */
     def markAsFailed(endpoint: ActorRef, timeOfRelease: Deadline): Unit =
       if (isWritable(endpoint)) {
-        addressToWritable += writableToAddress(endpoint) -> Gated(timeOfRelease)
-        writableToAddress -= endpoint
+        val address = writableToAddress(endpoint)
+        addressToWritable.get(address) match {
+          case Some(Quarantined(_, _)) ⇒ // don't overwrite Quarantined with Gated
+          case _ ⇒
```
Comment (MQ-EL) [inline akka-remote/src/main/scala/akka/remote/Remoting.scala]: Quarantine piercing lose quarantine marker (e.g. the state is ```Pass(_,_, Some(uid)```) ? and seems redundance when used with ```keepQuarantinedOr``` method together in the EM's ```supervisorStrategy```?

Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/Remoting.scala]: you mean that we can overwrite a `Pass` with some `refuseUid` with Gated and when that gating is removed the original `refuseUid` is lost?

Comment (MQ-EL) [inline akka-remote/src/main/scala/akka/remote/Remoting.scala]: If I understand your words correctly, not that, AFAIK. 
if the previous policy was a Pass with Some refusedUid, must to overwrite to quarantined. If it was already quarantined or Gated, need to update timestamp.

Comment (MQ-EL) [inline akka-remote/src/main/scala/akka/remote/Remoting.scala]: Perhaps better keep it unchanged, and add keepQuarantedOr logic in the caller's code？

Comment (MQ-EL) [inline akka-remote/src/main/scala/akka/remote/Remoting.scala]: correct (i am sorry):
>If it was already quarantined or Gated, ~~ok, that is impossible~~.

some race would happen due to async feature when actor stop.

**Comment by MQ-EL** [comment]
make an another test, but change node2 from pause to stop. They are the same in essence. you can find that the state from ```PASS``` to ```GATE``` losts refusedId. here are logs:

```log
[WARN] [05/26/2016 12:03:01.316] [ClusterSystem-akka.actor.default-dispatcher-16] [akka.tcp://ClusterSystem@127.0.0.1:2551/system/cluster/core/daemon] Cluster Node [akka.tcp://ClusterSystem@127.0.0.1:2551] - Marking node(s) as UNREACHABLE [Member(address = akka.tcp://ClusterSystem@127.0.0.1:2552, status = Up)]
[WARN] [05/26/2016 12:03:03.444] [ClusterSystem-akka.remote.default-remote-dispatcher-6] [akka.tcp://ClusterSystem@127.0.0.1:2551/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2FClusterSystem%40127.0.0.1%3A2552-1] Association with remote system [akka.tcp://ClusterSystem@127.0.0.1:2552] has failed, address is now gated for [5000] ms. Reason: [Association failed with [akka.tcp://ClusterSystem@127.0.0.1:2552]] Caused by: [Connection refused: no further information: /127.0.0.1:2552]
[INFO] [05/26/2016 12:03:03.448] [ClusterSystem-akka.remote.default-remote-dispatcher-6] [akka.remote.Remoting] ### Before state: [Some(Some(Pass(Actor[akka://ClusterSystem/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2FClusterSystem%40127.0.0.1%3A2552-1#-327271916],None,None)))]
[INFO] [05/26/2016 12:03:03.448] [ClusterSystem-akka.remote.default-remote-dispatcher-6] [akka.remote.Remoting] ### After state: [Some(Some(Gated(Deadline(271944194015957 nanoseconds))))]
[WARN] [05/26/2016 12:03:10.375] [ClusterSystem-akka.remote.default-remote-dispatcher-7] [akka.tcp://ClusterSystem@127.0.0.1:2551/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2FClusterSystem%40127.0.0.1%3A2552-2] Association with remote system [akka.tcp://ClusterSystem@127.0.0.1:2552] has failed, address is now gated for [5000] ms. Reason: [Association failed with [akka.tcp://ClusterSystem@127.0.0.1:2552]] Caused by: [Connection refused: no further information: /127.0.0.1:2552]
[INFO] [05/26/2016 12:03:10.376] [ClusterSystem-akka.remote.default-remote-dispatcher-7] [akka.remote.Remoting] ### Before state: [Some(Some(Pass(Actor[akka://ClusterSystem/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2FClusterSystem%40127.0.0.1%3A2552-2#-1052317757],None,None)))]
[INFO] [05/26/2016 12:03:10.376] [ClusterSystem-akka.remote.default-remote-dispatcher-7] [akka.remote.Remoting] ### After state: [Some(Some(Gated(Deadline(271951122484302 nanoseconds))))]
[INFO] [05/26/2016 12:03:11.336] [ClusterSystem-akka.actor.default-dispatcher-17] [akka.cluster.Cluster(akka://ClusterSystem)] Cluster Node [akka.tcp://ClusterSystem@127.0.0.1:2551] - Leader is auto-downing unreachable node [akka.tcp://ClusterSystem@127.0.0.1:2552]
[INFO] [05/26/2016 12:03:11.336] [ClusterSystem-akka.actor.default-dispatcher-18] [akka.cluster.Cluster(akka://ClusterSystem)] Cluster Node [akka.tcp://ClusterSystem@127.0.0.1:2551] - Marking unreachable node [akka.tcp://ClusterSystem@127.0.0.1:2552] as [Down]
[INFO] [05/26/2016 12:03:12.316] [ClusterSystem-akka.actor.default-dispatcher-4] [akka.cluster.Cluster(akka://ClusterSystem)] Cluster Node [akka.tcp://ClusterSystem@127.0.0.1:2551] - Leader is removing unreachable node [akka.tcp://ClusterSystem@127.0.0.1:2552]
[WARN] [05/26/2016 12:03:17.546] [ClusterSystem-akka.remote.default-remote-dispatcher-22] [akka.tcp://ClusterSystem@127.0.0.1:2551/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2FClusterSystem%40127.0.0.1%3A2552-3] Association with remote system [akka.tcp://ClusterSystem@127.0.0.1:2552] has failed, address is now gated for [5000] ms. Reason: [Association failed with [akka.tcp://ClusterSystem@127.0.0.1:2552]] Caused by: [Connection refused: no further information: /127.0.0.1:2552]
[INFO] [05/26/2016 12:03:17.547] [ClusterSystem-akka.remote.default-remote-dispatcher-22] [akka.remote.Remoting] ### Before state: [Some(Some(Pass(Actor[akka://ClusterSystem/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2FClusterSystem%40127.0.0.1%3A2552-3#1068937920],None,Some(1133361274))))]
[INFO] [05/26/2016 12:03:17.547] [ClusterSystem-akka.remote.default-remote-dispatcher-22] [akka.remote.Remoting] ### After state: [Some(Some(Gated(Deadline(271958293657950 nanoseconds))))]
[WARN] [05/26/2016 12:03:27.547] [ClusterSystem-akka.remote.default-remote-dispatcher-22] [akka.tcp://ClusterSystem@127.0.0.1:2551/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2FClusterSystem%40127.0.0.1%3A2552-4] Association with remote system [akka.tcp://ClusterSystem@127.0.0.1:2552] has failed, address is now gated for [5000] ms. Reason: [Association failed with [akka.tcp://ClusterSystem@127.0.0.1:2552]] Caused by: [Connection refused: no further information: /127.0.0.1:2552]
[INFO] [05/26/2016 12:03:27.547] [ClusterSystem-akka.remote.default-remote-dispatcher-25] [akka.remote.Remoting] ### Before state: [Some(Some(Pass(Actor[akka://ClusterSystem/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2FClusterSystem%40127.0.0.1%3A2552-4#-1380132429],None,None)))]
[INFO] [05/26/2016 12:03:27.547] [ClusterSystem-akka.remote.default-remote-dispatcher-25] [akka.remote.Remoting] ### After state: [Some(Some(Gated(Deadline(271968293376997 nanoseconds))))]
```
the ```EndpointManager```'s supervisorStrategy code snippet(the above logs start with "###" ): 

```scala
override val supervisorStrategy =
    OneForOneStrategy(loggingEnabled = false) {
      case e @ InvalidAssociation(localAddress, remoteAddress, reason, disassiciationInfo) ⇒
        ...
      case NonFatal(e) ⇒
        ...
        val addr = endpoints.addressForWriter(sender())
        log.info(s"### Before state: [${addr.map(endpoints.writableEndpointWithPolicyFor)}]")
        endpoints.markAsFailed(sender(), Deadline.now + settings.RetryGateClosedFor)
        log.info(s"### After state: [${addr.map(endpoints.writableEndpointWithPolicyFor)}]")
        ...
```

**Comment by patriknw** [comment]
@MQ-EL I'm not sure I understand all of your points, but I think I understand that the refuseUid can be lost when transitioning from `Pass` to `Gated`. I'll try to add `refuseUid` to `Gated` also so that we can carry it along for these transitions.

**Comment by patriknw** [comment]
@MQ-EL @drewhk pushed an attempt to keep the refuseUid for the transitions Pass -> Gated -> Pass

When reviewing you can think of the `WasGated` state as previously no entry in the endpoint registry, but it's no carrying the refuseUid.

**Comment by patriknw** [comment]
the failure was mima because of the changed classes, I will add filter

**Comment by MQ-EL** [comment]
I mean:https://github.com/MQ-EL/akka/commit/92f0038ba53386e108d77b7f3436504a1ae48794

and does the ```Gated``` state needed? if Pass1 -> Gate -> Pass2 happens, the Pass1's refusedId is out of date, because the Pass1's currentUid already exists.

**Comment by drewhk** [comment]
That is true, but nothing hurts when keeping around the old refuseUID?

**Comment by drewhk** [comment]
This look fine up to my understanding. We should go through the state transitions though in Berlin, it will be important for both the old and new remoting to nail this.

**Comment by patriknw** [comment]
Thanks for confirming.
In Artery, so far, it keeps track of all old quarantined UIDs, not only the latest.

**Comment by MQ-EL** [comment]
why keep all old quarantined UIDs in the Artery?

**Comment by patriknw** [comment]
We might revisit that, but I want to make it as safe and simple as possible. If it's quarantined we shall never talk to it again.

**Comment by patriknw** [comment]
@drewhk what do you think about @MQ-EL 's suggestion for publishAndThrow in https://github.com/MQ-EL/akka/commit/92f0038ba53386e108d77b7f3436504a1ae48794 ?

**Comment by drewhk** [comment]
WDYM? It sets it to None since the UID is unknown (handle.isEmpty). Am I missing somehting?

**Comment by drewhk** [comment]
The reason btw why we keep only one quarantine UID for an address because the assumption is that no two ActorSystems can exist on the same address+port, which means that a memory of one is enough. We can revisit this, just explaining the reasoning from back then.

**Comment by MQ-EL** [comment]
no @drewhk. ```keepQuarantinedOr``` will catch it :)

**Comment by MQ-EL** [comment]
see https://github.com/akka/akka/blob/master/akka-remote/src/main/scala/akka/remote/Remoting.scala#L489.

**Comment by patriknw** [comment]
I don't see that the change of `publishAndThrow` makes things easier to follow and if it's correct so I will not touch that here. @MQ-EL if you have a clear picture of something that is wrong I suggest that you create a separate PR, but perhaps it should not be solved in `publishAndThrow`.

**Comment by patriknw** [comment]
The test failure was same as https://github.com/akka/akka/issues/20676

**Comment by MQ-EL** [comment]
I see, thx.

**Pull request merged**